### ISSUE TITLE
[cpprestsdk] Build error. Building cpprestsdk:x64-windows failed with: BUILD_FAILED

### ISSUE THREAD

**Issue by gsaponaro-af** [microsoft/vcpkg#34745]
**Operating System**

Microsoft Windows Server 2022 10.0.203
from the [windows-2022](https://github.com/actions/runner-images/blob/win22/20231016.1/images/win/Windows2022-Readme.md) GitHub Actions GitHub-hosted runner.

**Compiler**

Visual Studio Enterprise 2022 17.7.34202.233
from the [windows-2022](https://github.com/actions/runner-images/blob/win22/20231016.1/images/win/Windows2022-Readme.md) GitHub Actions GitHub-hosted runner.

**Steps to reproduce the behavior**

```
vcpkg integrate install
vcpkg install cpprestsdk:x64-windows
```

**Failure logs**

<details>
  <summary>Click here</summary>

```
Run vcpkg integrate install
  vcpkg integrate install
  vcpkg install cpprestsdk:x64-windows
  shell: C:\Program Files\Git\bin\bash.EXE -l {0}
  env:
    BUILD_TYPE: Release
    INPUT_RUN_POST: true

Applied user-wide integration for this vcpkg root.
CMake projects should use: "-DCMAKE_TOOLCHAIN_FILE=C:/vcpkg/scripts/buildsystems/vcpkg.cmake"

All MSBuild C++ projects can now #include any installed libraries. Linking will be handled automatically. Installing new libraries will make them instantly available.
Computing installation plan...
The following packages will be built and installed:
  * brotli:x64-windows -> 1.1.0#1
    cpprestsdk[brotli,compression,core]:x64-windows -> 2.10.18#4
  * vcpkg-cmake:x64-windows -> 2023-05-04
  * vcpkg-cmake-config:x64-windows -> 2022-02-06#1
  * zlib:x64-windows -> 1.3
Additional packages (*) will be modified to complete this operation.
Detecting compiler hash for triplet x64-windows...
A suitable version of 7zip was not found (required v23.1.0) Downloading portable 7zip 23.1.0...
Downloading 7zip...
https://www.7-zip.org/a/7z2301-extra.7z->C:\vcpkg\downloads\7z2301-extra.7z
Downloading https://www.7-zip.org/a/7z2301-extra.7z
Extracting 7zip...
Restored 0 package(s) from C:\Users\runneradmin\AppData\Local\vcpkg\archives in 282 us. Use --debug to see more details.
Installing 1/5 vcpkg-cmake:x64-windows...
Building vcpkg-cmake:x64-windows...
-- Installing: C:/vcpkg/packages/vcpkg-cmake_x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake
-- Installing: C:/vcpkg/packages/vcpkg-cmake_x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake
-- Installing: C:/vcpkg/packages/vcpkg-cmake_x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake
-- Installing: C:/vcpkg/packages/vcpkg-cmake_x64-windows/share/vcpkg-cmake/vcpkg-port-config.cmake
-- Installing: C:/vcpkg/packages/vcpkg-cmake_x64-windows/share/vcpkg-cmake/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 88.4 ms.
Elapsed time to handle vcpkg-cmake:x64-windows: 187 ms
Installing 2/5 vcpkg-cmake-config:x64-windows...
Building vcpkg-cmake-config:x64-windows...
-- Installing: C:/vcpkg/packages/vcpkg-cmake-config_x64-windows/share/vcpkg-cmake-config/vcpkg_cmake_config_fixup.cmake
-- Installing: C:/vcpkg/packages/vcpkg-cmake-config_x64-windows/share/vcpkg-cmake-config/vcpkg-port-config.cmake
-- Installing: C:/vcpkg/packages/vcpkg-cmake-config_x64-windows/share/vcpkg-cmake-config/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 21.5 ms.
Elapsed time to handle vcpkg-cmake-config:x64-windows: 93.1 ms
Installing 3/5 brotli:x64-windows...
Building brotli:x64-windows...
-- Downloading https://github.com/google/brotli/archive/v1.1.0.tar.gz -> google-brotli-v1.1.0.tar.gz...
-- Extracting source C:/vcpkg/downloads/google-brotli-v1.1.0.tar.gz
-- Applying patch install.patch
-- Applying patch fix-arm-uwp.patch
-- Applying patch pkgconfig.patch
-- Applying patch emscripten.patch
-- Using source at C:/vcpkg/buildtrees/brotli/src/v1.1.0-958bece74c.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Fixing pkgconfig file: C:/vcpkg/packages/brotli_x64-windows/lib/pkgconfig/libbrotlicommon.pc
-- Fixing pkgconfig file: C:/vcpkg/packages/brotli_x64-windows/lib/pkgconfig/libbrotlidec.pc
-- Fixing pkgconfig file: C:/vcpkg/packages/brotli_x64-windows/lib/pkgconfig/libbrotlienc.pc
-- Downloading https://mirror.msys2.org/mingw/mingw32/mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst;https://repo.msys2.org/mingw/mingw32/mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/mingw/mingw32/mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/mingw/mingw32/mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/mingw/mingw32/mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst;https://mirror.selfnet.de/msys2/mingw/mingw32/mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst -> mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst...
-- Downloading https://mirror.msys2.org/msys/x86_64/msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst;https://repo.msys2.org/msys/x86_64/msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst;https://mirror.yandex.ru/mirrors/msys2/msys/x86_64/msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst;https://mirrors.tuna.tsinghua.edu.cn/msys2/msys/x86_64/msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst;https://mirrors.ustc.edu.cn/msys2/msys/x86_64/msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst;https://mirror.selfnet.de/msys2/msys/x86_64/msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst -> msys2-msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst...
-- Using msys root at C:/vcpkg/downloads/tools/msys2/6f3fa1a12ef85a6f
-- Fixing pkgconfig file: C:/vcpkg/packages/brotli_x64-windows/debug/lib/pkgconfig/libbrotlicommon.pc
-- Fixing pkgconfig file: C:/vcpkg/packages/brotli_x64-windows/debug/lib/pkgconfig/libbrotlidec.pc
-- Fixing pkgconfig file: C:/vcpkg/packages/brotli_x64-windows/debug/lib/pkgconfig/libbrotlienc.pc
-- Installing: C:/vcpkg/packages/brotli_x64-windows/share/brotli/usage
-- Installing: C:/vcpkg/packages/brotli_x64-windows/share/brotli/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 1.2 s.
Elapsed time to handle brotli:x64-windows: 54 s
Installing 4/5 zlib:x64-windows...
Building zlib:x64-windows...
-- Downloading https://github.com/madler/zlib/archive/v1.3.tar.gz -> madler-zlib-v1.3.tar.gz...
-- Extracting source C:/vcpkg/downloads/madler-zlib-v1.3.tar.gz
-- Applying patch 0001-Prevent-invalid-inclusions-when-HAVE_-is-set-to-0.patch
-- Applying patch 0002-skip-building-examples.patch
-- Applying patch 0003-build-static-or-shared-not-both.patch
-- Applying patch 0004-android-and-mingw-fixes.patch
-- Using source at C:/vcpkg/buildtrees/zlib/src/v1.3-8825ee792f.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
-- Building x64-windows-dbg
-- Building x64-windows-rel
-- Installing: C:/vcpkg/packages/zlib_x64-windows/share/zlib/vcpkg-cmake-wrapper.cmake
-- Fixing pkgconfig file: C:/vcpkg/packages/zlib_x64-windows/lib/pkgconfig/zlib.pc
-- Using cached mingw-w64-i686-pkgconf-1~1.8.0-2-any.pkg.tar.zst.
-- Using cached msys2-msys2-runtime-3.4.6-1-x86_64.pkg.tar.zst.
-- Using msys root at C:/vcpkg/downloads/tools/msys2/6f3fa1a12ef85a6f
-- Fixing pkgconfig file: C:/vcpkg/packages/zlib_x64-windows/debug/lib/pkgconfig/zlib.pc
-- Installing: C:/vcpkg/packages/zlib_x64-windows/share/zlib/copyright
-- Performing post-build validation
Stored binaries in 1 destinations in 135 ms.
Elapsed time to handle zlib:x64-windows: 28 s
Installing 5/5 cpprestsdk:x64-windows...
Building cpprestsdk[brotli,compression,core]:x64-windows...
-- Downloading https://github.com/Microsoft/cpprestsdk/archive/122d09549201da5383321d870bed45ecb9e168c5.tar.gz -> Microsoft-cpprestsdk-122d09549201da5383321d870bed45ecb9e168c5.tar.gz...
-- Extracting source C:/vcpkg/downloads/Microsoft-cpprestsdk-122d09549201da5383321d870bed45ecb9e168c5.tar.gz
-- Applying patch fix-find-openssl.patch
-- Applying patch fix_narrowing.patch
-- Applying patch fix-uwp.patch
-- Using source at C:/vcpkg/buildtrees/cpprestsdk/src/ecb9e168c5-295a1e905e.clean
-- Found external ninja('1.11.0').
-- Configuring x64-windows
CMake Warning at installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_configure.cmake:318 (message):
  The following variables are not used in CMakeLists.txt:

      WEBSOCKETPP_CONFIG
      WEBSOCKETPP_CONFIG_VERSION

  Please recheck them and remove the unnecessary options from the
  `vcpkg_cmake_configure` call.

  If these options should still be passed for whatever reason, please use the
  `MAYBE_UNUSED_VARIABLES` argument.
Call Stack (most recent call first):
  ports/cpprestsdk/portfile.cmake:33 (vcpkg_cmake_configure)
  scripts/ports.cmake:168 (include)


-- Building x64-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:134 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j3
    Working Directory: C:/vcpkg/buildtrees/cpprestsdk/x64-windows-dbg
    See logs for more information:
      C:\vcpkg\buildtrees\cpprestsdk\install-x64-windows-dbg-out.log

Call Stack (most recent call first):
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_build.cmake:74 (vcpkg_execute_build_process)
  installed/x64-windows/share/vcpkg-cmake/vcpkg_cmake_install.cmake:16 (vcpkg_cmake_build)
  ports/cpprestsdk/portfile.cmake:[48](https://github.com/redacted/redacted/actions/runs/redacted/job/redacted#step:10:49) (vcpkg_cmake_install)
  scripts/ports.cmake:168 (include)


error: building cpprestsdk:x[64](https://github.com/redacted/redacted/actions/runs/redacted/job/redacted#step:10:65)-windows failed with: BUILD_FAILED
Elapsed time to handle cpprestsdk:x64-windows: 31 s
Please ensure you're using the latest port files with `git pull` and `vcpkg update`.
Then check for known issues at:
  https://github.com/microsoft/vcpkg/issues?q=is%3Aissue+is%3Aopen+in%3Atitle+cpprestsdk
You can submit a new issue at:
  https://github.com/microsoft/vcpkg/issues/new?title=[cpprestsdk]+Build+error&body=Copy+issue+body+from+C%3A%2Fvcpkg%2Finstalled%2Fvcpkg%2Fissue_body.md
You can also submit an issue by running (GitHub CLI must be installed):
  gh issue create -R microsoft/vcpkg --title "[cpprestsdk] Build failure" --body-file C:/vcpkg/installed/vcpkg/issue_body.md
Error: Process completed with exit code 1.
```

</details>

**Additional context**

I tried restarting the GitHub Actions job three times. It fails consistently.

**Comment by MonicaLiu0311** [comment]
> ```
>     See logs for more information:
>       C:\vcpkg\buildtrees\cpprestsdk\install-x64-windows-dbg-out.log
> ```

Please provide detailed logs.

**Comment by gsaponaro-af** [comment]
> > ```
> >     See logs for more information:
> >       C:\vcpkg\buildtrees\cpprestsdk\install-x64-windows-dbg-out.log
> > ```
> 
> Please provide detailed logs.

Here it is: https://gist.github.com/gsaponaro-af/b08f240b6d47d80efc9e329ee17406b9

Thank you

**Comment by MonicaLiu0311** [comment]
Do you fall into one of the following situations:

1. If you have a static library checked-in to your project which was built before `Visual Studio 2022 version 17.2 Preview 2` and you build the rest of your project with a newer toolset you will see this error. 
    See: https://developercommunity.visualstudio.com/t/__imp___std_init_once_complete-unresolve/1684365#T-N10047968

    The fix has been checked in for `17.3 Preview 3`. 
See: https://developercommunity.visualstudio.com/t/__imp___std_init_once_complete-unresolve/1684365#T-N10053839
    
2. You’re compiling some object files/static libraries with `VS 2022`, but performing the final link with `VS 2019`. 
    See: 
    https://developercommunity.visualstudio.com/t/__imp___std_init_once_complete-unresolve/1684365#T-N10079675
    https://developercommunity.visualstudio.com/t/__imp___std_init_once_complete-unresolve/1684365#T-N10080626

    If you have a different version of VS installed, the VS that vcpkg links to may not be the one you expect. You can specify this in `VCPKG_PATH/triplets/install_triplet.cmake`. 
    See: https://github.com/microsoft/vcpkg/issues/23308

**Comment by gsaponaro-af** [comment]
Good news, this issue can be closed.

My GitHub runner job was structured like this:

1. Set up a Conda environment via the GitHub Action [conda-incubator/setup-miniconda](https://github.com/conda-incubator/setup-miniconda)
2. Install C++ REST SDK using `vcpkg install cpprestsdk:x64-windows`
3. Set up Visual Studio shell via the GitHub Action [egor-tensin/vs-shell](https://github.com/egor-tensin/vs-shell)

Now, I inverted the order of the first two steps, therefore placing the vcpkg step first, and the issue is gone!

I am convinced that the cause was "compiling some object files/static libraries with VS 2022, but performing the final link with VS 2019". I did not investigate further though. Thank you for the support.
